### PR TITLE
Display redeemers in `transaction view`

### PR DESCRIPTION
**Pull request by palas**
# Changelog

```yaml
- description: |
    Add info about redeemers to output of `transaction view` command.
  type:
  - feature        # introduces a new feature
```

# Context

It increases the amount of info displayed. It is only applicable to Shelley and after. It uses `cborg-json` to render the CBOR of the redeemer.

# How to trust this PR

I would check:
 - That the approach makes sense overall.
 - That it makes sense to return `null` when it does and it looks right in the changed tests.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [x] Self-reviewed the diff

### CODE REVIEW

**Comment by smelc** [comment]
LGTM but I'm not approving since I never manipulated nor cbor nor redeemers :laughing:

File: cardano-cli/test/cardano-cli-golden/files/golden/alonzo/transaction-view.out [hunk 1]
```diff
@@ -9,6 +9,7 @@ inputs:
 metadata: null
 mint: null
 outputs: []
+redeemers: []
```
Comment (carbolymer) [inline cardano-cli/test/cardano-cli-golden/files/golden/alonzo/transaction-view.out:12]: Can you add a test case for non-empty redeemers?

File: cardano-cli/src/Cardano/CLI/Json/Friendly.hs [hunk 2]
```diff
@@ -172,7 +176,16 @@ friendlyTxBodyImpl
...
+    ])
+
+redeemerIfShelleyBased :: CardanoEra era -> TxBody era -> [Aeson.Pair]
+redeemerIfShelleyBased era tb = caseByronOrShelleyBasedEra [] (\shEra -> [ "redeemers" .= friendlyRedeemer shEra tb ] ) era
+
+friendlyRedeemer :: ShelleyBasedEra era -> TxBody era -> Aeson.Value
+friendlyRedeemer _ (ShelleyTxBody _ _ _ TxBodyNoScriptData _ _) = Aeson.Null
+friendlyRedeemer _ (ShelleyTxBody _ _ _ (TxBodyScriptData _ _ r) _ _) = encodingToJSON $ L.toCBOR r
+  where encodingToJSON :: Encoding -> Aeson.Value
+        encodingToJSON e = either (Aeson.String . Text.pack) id $ fromFlatTerm (decodeValue True) $ toFlatTerm e
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/Json/Friendly.hs]: I'm not sure about this failure here - instead of an error, user will see an error text in the redeemers field. That may be confusing.

Have you tried to autogenerate `ToJSON` for `Redeemers`? [It has `Generic`](https://cardano-ledger.cardano.intersectmbo.org/cardano-ledger-alonzo/Cardano-Ledger-Alonzo-TxWits.html#t:Redeemers) instance so maybe simply an orphan
```haskell
instance ToJSON Redeemers
```
could work here?

Comment (palas) [reply cardano-cli/src/Cardano/CLI/Json/Friendly.hs]: Hm, after a lot of trying, that doesn't seem to work. I can use `genericToJSON` to make up for `Redeemers` not having `ToJSON` but then `RedeemersRaw` doesn't have it either, so I don't see any easy way to do the conversion from `cardano-cli`.  What may work would be to go to `cardano-ledger` and add the `ToJSON` instances there.

Comment (palas) [reply cardano-cli/src/Cardano/CLI/Json/Friendly.hs]: Other possible approaches (alternatives to adding `ToJSON` to `cardano-ledger`):
- Crash when `decodeValue` crashes
- Don't return it when `decodeValue` crashes
- Implement our own `decodeValue` (or fork it) and make it exhaustive
- Returning the result of showing the `Encoding` type
- Having a sum type in the JSON: something like `{decodedRedeemer = {...}}` and having `{error = "Could not decode because blah..."}`

File: cardano-cli/test/cardano-cli-golden/files/golden/alonzo/transaction-view.out (hunk 1 above)
Comment (palas) [reply cardano-cli/test/cardano-cli-golden/files/golden/alonzo/transaction-view.out:12]: Added in [this commit](https://github.com/IntersectMBO/cardano-cli/pull/664/commits/fb6c1b258f9ecf7e11b0fe21526d58a8a6a01b7c) :+1:

File: cardano-cli/src/Cardano/CLI/Json/Friendly.hs (hunk 2 above)
Comment (carbolymer) [reply cardano-cli/src/Cardano/CLI/Json/Friendly.hs]: You're right, it's not easy to generate automatic `ToJSON` here. Thank you for checking that. I think your solution is the best for what I can think of. You can also ask on #ledger if they know if there's a better way to print `Redeemers`.

With regards to the failure reporting, I think crashing here would be too drastic. What do you think about printing the error to `stderr` and setting the field to null value?
(You'd need to change the return type to include the possible error besides the JSON value).

Comment (palas) [reply cardano-cli/src/Cardano/CLI/Json/Friendly.hs]: I see, because it is `stderr` it won't mess with `stdout`. I will give that a try

Comment (palas) [reply cardano-cli/src/Cardano/CLI/Json/Friendly.hs]: Ok, done in [this Commit](https://github.com/IntersectMBO/cardano-cli/pull/664/commits/dda49bd1213024091cba1e699b12aa3ae5578f78) 😎. I had a bit of fun, but I documented everything as well as I could.

File: cardano-cli/src/Cardano/CLI/Types/MonadWarning.hs
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/Types/MonadWarning.hs:1]: Can you reindent the file please? It should be 2 spaces instead of 4.

File: cardano-cli/src/Cardano/CLI/Types/MonadWarning.hs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+{-# LANGUAGE InstanceSigs #-}
+
+module Cardano.CLI.Types.MonadWarning
+  ( MonadWarning(..)
+  , eitherToWarning
+  , monadWarningInMonadIO
+  , monadWarningInStateT
+  ) where
+
+import           Cardano.Prelude (MonadIO (liftIO), MonadState (state), Print (hPutStrLn), StateT,
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/Types/MonadWarning.hs]: I think the direction is to move away from cardano-prelude. Can you use imports from the original packages instead?

File: cardano-cli/src/Cardano/CLI/Types/MonadWarning.hs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    deriving (Functor, Applicative, Monad, MonadIO)
+
+-- | This instance prints the issue to the 'stderr'.
+instance MonadIO m => MonadWarning (WarningMonadIO m) where
+    reportIssue :: String -> WarningMonadIO m ()
+    reportIssue issue = liftIO (hPutStrLn stderr issue)
+
+-- | Convert a 'MonadWarning' into a 'MonadIO' by reporting
+-- warnings to 'stderr'.
+monadWarningInMonadIO :: WarningMonadIO m a -> m a
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/Types/MonadWarning.hs]: The standard in mtl seems to call such functions using a verb `runXXX` e.g. `runExcept`. I think just
```haskell
newtype WarningIO m a = WarningIO { runWarningIO :: m a }
    deriving (Functor, Applicative, Monad, MonadIO)

newtype WarningStateT m a = WarningStateT { runWarningStateT :: StateT [String] m a }
    deriving (Functor, Applicative, Monad, MonadState [String])
```
without this and `monadWarningInStateT` functions would be enough. Also, the name `Monad` in the newtypes seems redundant.

**Review by carbolymer — APPROVED** [review]

File: cardano-cli/src/Cardano/CLI/Types/MonadWarning.hs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  reportIssue :: String -- ^ The warning message to report.
+              -> m ()   -- ^ The action that reports the warning.
+
+-- | Wrapper newtype for 'MonadIO' with 'MonadWarning' instance.
+-- We need to have wrapper to avoid overlapping instances.
+newtype WarningIO m a = WarningIO { runWarningIO :: m a }
+  deriving (Functor, Applicative, Monad, MonadIO)
+
+-- | This instance prints the issue to the 'stderr'.
+instance MonadIO m => MonadWarning (WarningIO m) where
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Types/MonadWarning.hs:65]: Nice, you implemented this type class so when printing a tx, is something fails we don't fail catastrophically.

**Review by Jimbo4350 — COMMENTED** [review]
:+1:

File: cardano-cli/src/Cardano/CLI/Types/MonadWarning.hs (hunk 5 above)
Comment (palas) [reply cardano-cli/src/Cardano/CLI/Types/MonadWarning.hs:65]: Yes, exactly, in response to [this review](https://github.com/IntersectMBO/cardano-cli/pull/664#discussion_r1535313731) by @carbolymer
